File: sdk/dotnet/Shojiku/LocalPem.cs
// A signing provider backed by a PEM key and certificate.
//
// The only provider this release has. KMS and HSM providers are a recorded
// deferral, which is why this is a named class rather than a pair of arguments
// on `Sign` — a second provider then adds a class, not a signature change in
// seven languages.
//
// The material comes either from paths (`key` / `cert`) or from bytes already in
// memory (`keyPem` / `certPem`), so a key fetched from a secret manager never
// has to be written to disk first. Which one you passed is explicit rather than
// sniffed: guessing whether a string is a path or a PEM body is exactly the kind
// of cleverness that reads the wrong file.
//
// Nothing here logs key material, and the engine builds its refusals from fixed
// strings, so a rejection cannot echo it back either.

namespace Shojiku;

/// <summary>What a client needs from anything it can sign with.</summary>
[System.Diagnostics.CodeAnalysis.SuppressMessage(
    "Design",
    "CA1040:Avoid empty interfaces",
    Justification =
        "A marker is what this IS: the operation a provider performs crosses internal types, " +
        "so the hook lives on the internal IEngineSigner and this names the public concept a " +
        "caller passes to Sign. Giving it a member would put an engine type on the public surface.")]
public interface ISigningProvider
{
}

/// <summary>What a client actually needs from a provider: a signed document.</summary>
/// <remarks>
/// The polymorphic hook, so <see cref="ShojikuClient.Sign"/> branches on
/// nothing — what differs between a key held in this process and one held in a
/// cloud service is HOW a signature is produced, which is exactly what this
/// method is. It is INTERNAL because it crosses internal types (the engine and
/// its snapshot), which also makes <see cref="ISigningProvider"/> a marker: the
/// providers this package ships are the implementations, and a provider of your
/// own is a design change rather than a subclass.
/// </remarks>
internal interface IEngineSigner : ISigningProvider
{
    Snapshot SignWith(Engine engine, byte[] pdf);
}

/// <summary>PEM key + certificate, from paths or from bytes, never sniffed.</summary>
public sealed class LocalPem : IEngineSigner
{
    private readonly string? keyPath;
    private readonly string? certPath;
    private byte[]? keyPem;
    private byte[]? certPem;

    /// <summary>Creates a provider from any one form of each half.</summary>
    /// <param name="key">Path to the private key.</param>
    /// <param name="cert">Path to the certificate.</param>
    /// <param name="keyPem">The private key as bytes.</param>
    /// <param name="certPem">The certificate as bytes.</param>
    /// <param name="passphrase">The key's passphrase, when it has one.</param>
    public LocalPem(
        string? key = null,
        string? cert = null,
        byte[]? keyPem = null,
        byte[]? certPem = null,
        byte[]? passphrase = null)
    {
        keyPath = key;
        certPath = cert;
        this.keyPem = keyPem;
        this.certPem = certPem;
        Passphrase = passphrase;
        OneSource(key, keyPem, "key");
        OneSource(cert, certPem, "cert");
    }

    /// <summary>The private key, as PEM or DER bytes.</summary>
    public byte[] Key => keyPem ??= Text.ReadMaterial(keyPath!, "key_unreadable");

    /// <summary>The signing certificate, as PEM or DER bytes.</summary>
    public byte[] Certificate => certPem ??= Text.ReadMaterial(certPath!, "certificate_unreadable");

    /// <summary>The key's passphrase, when it has one.</summary>
    public byte[]? Passphrase { get; }

    /// <summary>Signs with the key this process holds.</summary>
    Snapshot IEngineSigner.SignWith(Engine engine, byte[] pdf) =>
        engine.Sign(pdf, Key, Certificate, Passphrase);

    /// <summary>
    /// Redacted, deliberately.
    /// </summary>
    /// <remarks>
    /// The default <c>ToString</c> would be the type name, but an object
    /// initializer, a structured logger or a debugger display reaches for the
    /// members — which here are the private key and the passphrase. So this
    /// states what is safe and nothing else: the class, and which FORM each
    /// half came from. Registering the provider once shrinks the surface
    /// further: material loads into one object instead of being rebuilt per
    /// request.
    /// </remarks>
    public override string ToString() =>
        $"<LocalPem key={Form(keyPath)} cert={Form(certPath)} passphrase={(Passphrase is null ? "none" : "[redacted]")}>";

    /// <summary>
    /// The path, or a note that the bytes came from memory.
    /// </summary>
    /// <remarks>
    /// A configured file path is not secret and is the one thing worth seeing
    /// when a provider loaded the wrong material; the bytes themselves are
    /// never printed.
    /// </remarks>
    private static string Form(string? path) => path ?? "[pem bytes]";

    /// <summary>
    /// Explicit, never sniffed — in BOTH directions.
    /// </summary>
    /// <remarks>
    /// Guessing whether a string is a path or a PEM body is how the wrong file
    /// gets read; accepting both forms and silently preferring one ignores the
    /// argument the caller meant, which is the same mistake one layer quieter.
    /// </remarks>
    private static void OneSource(string? path, byte[]? pem, string what)
    {
        var forms = $"`{what}` (a path) or `{what}Pem` (bytes)";
        if (path is not null && pem is not null)
        {
            throw new UsageException($"LocalPem takes either {forms}, not both");
        }

        if (path is null && pem is null)
        {
            throw new UsageException($"LocalPem needs either {forms}");
        }
    }
}

File: sdk/java/src/main/java/jp/kengos/shojiku/LocalPem.java
package jp.kengos.shojiku;

import java.nio.charset.StandardCharsets;
import java.nio.file.Path;

/**
 * A signing provider backed by a PEM key and certificate.
 *
 * <p>The only provider this release has. KMS and HSM providers are a recorded deferral, which is
 * why this is a named class rather than a pair of arguments on {@code sign} — a second provider
 * then adds a class, not a signature change in seven languages.
 *
 * <p>The material comes either from paths or from bytes already in memory, so a key fetched from a
 * secret manager never has to be written to disk first. Which one you passed is explicit rather
 * than sniffed: guessing whether a string is a path or a PEM body is exactly the kind of cleverness
 * that reads the wrong file. Java's types make that explicit at the call site — a {@link Path} is a
 * path and a {@code byte[]} is material — and the builder still refuses BOTH forms of the same
 * half.
 *
 * <p>Nothing here logs key material, and the engine builds its refusals from fixed strings, so a
 * rejection cannot echo it back either.
 */
public final class LocalPem implements EngineSigner {

  private final Path keyPath;
  private final Path certPath;
  private final byte[] passphrase;
  private byte[] keyPem;
  private byte[] certPem;

  private LocalPem(Builder builder) {
    this.keyPath = builder.keyPath;
    this.certPath = builder.certPath;
    this.keyPem = builder.keyPem;
    this.certPem = builder.certPem;
    this.passphrase = builder.passphrase;
    oneSource(builder.keyPath, builder.keyPem, "key");
    oneSource(builder.certPath, builder.certPem, "cert");
  }

  /**
   * The common case: both halves as files on disk.
   *
   * @param key path to the private key
   * @param cert path to the certificate
   */
  public LocalPem(Path key, Path cert) {
    this(builder().key(key).cert(cert));
  }

  /**
   * A builder, for the mixed and in-memory forms.
   *
   * @return a new builder
   */
  public static Builder builder() {
    return new Builder();
  }

  /**
   * The private key, as PEM or DER bytes.
   *
   * @return the key material
   */
  public byte[] key() {
    if (keyPem == null) {
      keyPem = Texts.readMaterial(keyPath, "key_unreadable");
    }
    return keyPem;
  }

  /**
   * The signing certificate, as PEM or DER bytes.
   *
   * @return the certificate material
   */
  public byte[] certificate() {
    if (certPem == null) {
      certPem = Texts.readMaterial(certPath, "certificate_unreadable");
    }
    return certPem;
  }

  /**
   * The key's passphrase, when it has one.
   *
   * @return the passphrase, or null
   */
  public byte[] passphrase() {
    return passphrase;
  }

  /**
   * Redacted, deliberately.
   *
   * <p>The default {@code toString} would be a class name and a hash, but a debugger's display, a
   * structured logger or any message that interpolates the provider reaches for the members — which
   * here are the private key and the passphrase. So this states what is safe and nothing else: the
   * class, and which FORM each half came from. Registering the provider once shrinks the surface
   * further: material loads into one object instead of being rebuilt per request.
   *
   * @return the redacted form
   */
  @Override
  public Snapshot signWith(Engine engine, byte[] pdf) {
    return engine.sign(pdf, key(), certificate(), passphrase());
  }

  @Override
  public String toString() {
    return "<LocalPem key="
        + form(keyPath)
        + " cert="
        + form(certPath)
        + " passphrase="
        + (passphrase == null ? "none" : "[redacted]")
        + ">";
  }

  /**
   * The path, or a note that the bytes came from memory.
   *
   * <p>A configured file path is not secret and is the one thing worth seeing when a provider
   * loaded the wrong material; the bytes themselves are never printed.
   */
  private static String form(Path path) {
    return path == null ? "[pem bytes]" : path.toString();
  }

  /**
   * Explicit, never sniffed — in BOTH directions.
   *
   * <p>Accepting both forms and silently preferring one ignores the argument the caller meant, on
   * the path where reading the wrong key matters most.
   */
  private static void oneSource(Path path, byte[] pem, String what) {
    String forms = "`" + what + "` (a path) or `" + what + "Pem` (bytes)";
    if (path != null && pem != null) {
      throw new UsageException("LocalPem takes either " + forms + ", not both");
    }
    if (path == null && pem == null) {
      throw new UsageException("LocalPem needs either " + forms);
    }
  }

  /** Builds a {@link LocalPem} from any one form of each half. */
  public static final class Builder {
    private Path keyPath;
    private Path certPath;
    private byte[] keyPem;
    private byte[] certPem;
    private byte[] passphrase;

    private Builder() {}

    /**
     * The private key, as a file.
     *
     * @param path where the key is
     * @return this builder
     */
    public Builder key(Path path) {
      this.keyPath = path;
      return this;
    }

    /**
     * The certificate, as a file.
     *
     * @param path where the certificate is
     * @return this builder
     */
    public Builder cert(Path path) {
      this.certPath = path;
      return this;
    }

    /**
     * The private key, as bytes already in memory.
     *
     * @param pem the key material
     * @return this builder
     */
    public Builder keyPem(byte[] pem) {
      this.keyPem = pem == null ? null : pem.clone();
      return this;
    }

    /**
     * The certificate, as bytes already in memory.
     *
     * @param pem the certificate material
     * @return this builder
     */
    public Builder certPem(byte[] pem) {
      this.certPem = pem == null ? null : pem.clone();
      return this;
    }

    /**
     * The key's passphrase, when it has one.
     *
     * @param value the passphrase
     * @return this builder
     */
    public Builder passphrase(byte[] value) {
      this.passphrase = value == null ? null : value.clone();
      return this;
    }

    /**
     * The key's passphrase as text, encoded UTF-8.
     *
     * @param value the passphrase
     * @return this builder
     */
    public Builder passphrase(String value) {
      return passphrase(value == null ? null : value.getBytes(StandardCharsets.UTF_8));
    }

    /**
     * Builds the provider, refusing both forms of either half.
     *
     * @return the provider
     */
    public LocalPem build() {
      return new LocalPem(this);
    }
  }
}
